db/wide/wide_column_serialization.cc

```diff
@@ -0,0 +1,162 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#include "db/wide/wide_column_serialization.h"
+
+#include <algorithm>
+#include <cassert>
+
+#include "port/port.h"
+#include "util/autovector.h"
+#include "util/coding.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+Status WideColumnSerialization::Serialize(const WideColumnDescs& column_descs,
+                                          std::string* output) {
+  // Column names should be strictly ascending
+  assert(std::adjacent_find(
+             column_descs.cbegin(), column_descs.cend(),
+             [](const WideColumnDesc& lhs, const WideColumnDesc& rhs) {
+               return lhs.name().compare(rhs.name()) > 0;
+             }) == column_descs.cend());
+  assert(output);
+
+  if (column_descs.size() > static_cast<size_t>(port::kMaxUint32)) {
+    return Status::InvalidArgument("Too many wide columns");
+  }
+
+  PutVarint32(output, kCurrentVersion);
+
+  PutVarint32(output, static_cast<uint32_t>(column_descs.size()));
+
+  size_t total_column_value_size = 0;
+
+  for (const auto& desc : column_descs) {
+    const Slice& name = desc.name();
+    if (name.size() > static_cast<size_t>(port::kMaxUint32)) {
+      return Status::InvalidArgument("Wide column name too long");
+    }
+
+    const Slice& value = desc.value();
+    if (value.size() > static_cast<size_t>(port::kMaxUint32)) {
+      return Status::InvalidArgument("Wide column value too long");
+    }
+
+    PutLengthPrefixedSlice(output, name);
+    PutVarint32(output, static_cast<uint32_t>(value.size()));
+
+    total_column_value_size += value.size();
+  }
+
+  output->reserve(output->size() + total_column_value_size);
+
+  for (const auto& desc : column_descs) {
+    const Slice& value = desc.value();
+
+    output->append(value.data(), value.size());
+  }
+
+  return Status::OK();
+}
+
+Status WideColumnSerialization::DeserializeOne(Slice* input,
+                                               const Slice& column_name,
+                                               WideColumnDesc* column_desc) {
```

Comment:
Deserializing everything and then binary search the column name is obviously not the most efficient way of deserializing one column. Is it a temporary solution?

Reply:
Because we use variable-length encoding in this format, we actually have to deserialize the "index" part completely in order to find out where the index ends and the column values start. If this is an issue, we can store the start offset of the column values as an extra (fixed-size encoded) footer. Since we have versioning in the format, I feel we can always add this later but let me know what you think.